src/iterators/base-iterator.ts

```diff
@@ -0,0 +1,40 @@
+export const DONE = { done: true, value: undefined };
+
+const KEEP_ALIVE = Symbol('keepAlive');
+
+/**
+ * The base class for iterators.
+ */
+export class BaseIterator<T> implements Iterator<T>, AsyncIterator<T> {
+	protected finished = false;
+	protected iterator: Iterator<T> | AsyncIterator<T>;
+
+	constructor(iterator: Iterator<T> | AsyncIterator<T>, keepAlive = false) {
+		this.iterator = iterator;
+		if (keepAlive) {
+			this[KEEP_ALIVE] = true;
+			iterator[KEEP_ALIVE] = true;
+		}
+	}
+
+	next(): IteratorResult<T> | Promise<IteratorResult<T>> | any {
```

Comment:
I don't like the `| any` - i think that basically neutralizes the other types you've listed here. But I'll have to follow up after I run this locally and inspect the type system a bit closer

Reply:
I concur. I need the `any` to make TypeScript happy because this is both an `Iterator` and `AsyncIterator`. JavaScript has no problem with this, but in TypeScript, `Iterator` and `AsyncIterator` conflict. You're not supposed to do both. The best solution I could come up with is `any`.